### PR TITLE
[RFC][inductor] Add collective function fusion: AllReduce + Split to ReduceScatter

### PR DESCRIPTION
**Pull request by yoyoyocmu**
Summary:
In this change, we add the support to fuse collective communication function that matches following pattern:
```
From:
    AllReduce() --> wait() --> split() --> getitem()

To:
    ReduceScatter --> wait()
```

Test Plan:
**CI Test**:
- CI link:

**Local Test**:
This is the trace before vs. after:
https://pxl.cl/4m9Lw 
https://pxl.cl/4m9Mr

Differential Revision: D53770216

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov

### CODE REVIEW

File: torch/_inductor/fx_passes/comm_fusion.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    if split_dim != 0:
+        return
+
+    split_size = split_node.args[1]
+    all_reduce_shape = node.meta["val"].shape[split_dim]
+    if split_size * len(node.args[3]) != all_reduce_shape:
+        return
+
+    # Make sure the getitem index is the same as the rank idx
+    current_rank = torch.distributed.get_rank()
```
Comment (wconstab) [inline torch/_inductor/fx_passes/comm_fusion.py:89]: Should we actually find the pg used for the allreduce op and then use the local rank from that pg for matching and constructing the new reducescatter?

File: torch/_inductor/config.py [hunk 2]
```diff
@@ -168,6 +168,11 @@ def is_fbcode():
     "raise_comms",
 ]
 
+# enable collective communication function fusion pass (Experimental)
+# e.g., fuse allreduce + split into a single reducescatter)
+# (if pattern_matcher is False, the enable_comm_fusion_pass flag will be ignored)
+enable_comm_fusion_pass = False
```
Comment (wconstab) [inline torch/_inductor/config.py:174]: We'll probably add more fusions. Are we going to want to turn them on independently of each other? Make this config accept a list of fusions to run, or just keep it book but name it for allreduce_split fusion. @yf225 what do you think

**Review by wconstab — COMMENTED** [review]
Looks good to me! Just a couple questions

File: torch/_inductor/config.py (hunk 2 above)
Comment (fegin) [reply torch/_inductor/config.py:174]: I like the idea of what @yf225 did for reordering. The reordering optimization take a list of string and apply the passes in the list. Users can also add customized pass by adding a function pointer to the list.

File: torch/_inductor/fx_passes/comm_fusion.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+comm_fusion_patterns = PatternMatcherPass()
+
+
+def comm_fusion_passes(graph: torch.fx.Graph) -> None:
+    """Pattern matching passes for collective communication fucntion fusion."""
+    if not torch.distributed.is_available:
```
Comment (fegin) [inline torch/_inductor/fx_passes/comm_fusion.py:25]: We should check the distributed availability in `post_grad.py` and assume distributed is available inside this module.

File: torch/_inductor/fx_passes/comm_fusion.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            torch.ops.c10d_functional.reduce_scatter_tensor.default,
+            args=(
+                node.args[0],  # input
+                node.args[1],  # reduce_op
+                node.args[2],  # tag
+                node.args[3],  # ranks
+                node.args[4],  # group_size
+            ),
+        )
+        new_wait_node = graph.call_function(
```
Comment (fegin) [inline torch/_inductor/fx_passes/comm_fusion.py:113]: The wait node will be right after the reducer_scatter, is no comm/comp overlapping expected?

File: torch/_inductor/fx_passes/comm_fusion.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    if wait_node is None:
+        return
+
+    split_node = CommFusionHelper.get_only_child(
+        wait_node, "call_function", torch.ops.aten.split.Tensor
+    )
+    if split_node is None:
+        return
+
+    getitem_node = CommFusionHelper.get_only_child(
```
Comment (bdhirsh) [inline torch/_inductor/fx_passes/comm_fusion.py:68]: Hmm - instead of doing this all manually, I think it should be possible to specify the entire pattern in the decorator? Something like this (I may have gotten some parts wrong - mostly went off some examples from [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/post_grad.py#L305)):
```
@register_graph_pattern(
    CallFunction(
        aten.split.Tensor,
        CallFunction(
           c10d_functional.wait_tensor.default,
           CallFunction(
              c10d_functional.all_reduce.default,users=MULTIPLE,
           ),
        ),
        dim=0,
        users=MULTIPLE,
    ),
```

File: test/distributed/test_inductor_collectives.py [hunk 6]
```diff
@@ -121,6 +121,42 @@ def compile(func, example_inputs):
...
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(2)
+    # TODO: somehow inductor bg compile threads are causing hangs at exit with distributed work dtor
+    @patch.object(torch._inductor.config, "compile_threads", 1)
+    @patch.object(torch._inductor.config, "enable_comm_fusion_pass", True)
+    def test_comm_fusion_inductor(self):
+        """Test that inductor fx pass fuses comm_fusion patterns."""
+
+        def allreduce_split(a, idx, split_size, *, tag, ranks, group_size):
+            ar = torch.ops.c10d_functional.all_reduce(a, "sum", tag, ranks, group_size)
```
Comment (bdhirsh) [inline test/distributed/test_inductor_collectives.py:133]: Maybe worth having some tests for when we don't match successfully?

(1) more than one output from the `aten.split()` calls is consumed by a downstream op
(2) split index is not equal to current local rank

File: torch/_inductor/config.py (hunk 2 above)
Comment (fegin) [reply torch/_inductor/config.py:174]: Since it is unclear how we would group and apply the future distributed optimization passes, I suggest changing the flag naming atm because the current naming is too general as @wconstab also mentioned. ,  @yf225

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 4 above)
Comment (fegin) [reply torch/_inductor/fx_passes/comm_fusion.py:113]: That's true. But I think the reducer_scatter should be created right after the original all_reduce node instead of being created before the wait node. This would allow us to preserve the original collective node and wait distance in the graph. If the reordering pass is not enabled but the original graph has some overlapping, we can still preserve the overlapping.

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 3 above)
Comment (yoyoyocmu) [reply torch/_inductor/fx_passes/comm_fusion.py:25]: Appreciate the suggestion, I think placing the distributed availability check within the communication fusion-only file helps keep a more self-contained structure. Given that the post_grad file is designed to be general, I prefer not to introduce this check there to avoid potential confusion in that broader context.

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 1 above)
Comment (yoyoyocmu) [reply torch/_inductor/fx_passes/comm_fusion.py:89]: Yeah, this is what we are doing now:

1. pg or rank_list is first extracted from the allreduce `args[3]`;
2. we then get the local rank and check if it is already in the above rank_list, and if the index match the following getitem() input;
3. after passing the check we reuse the pg from step 1 and construct the new reducescatter.

Let me add more context in the comment to make it clear.

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 4 above)
Comment (yoyoyocmu) [reply torch/_inductor/fx_passes/comm_fusion.py:113]: `wait()` reordering happens after this post_grad stage and in inductor scheduler. Here we actually remove allreduce+wait+split and replace it with reducescatter+wait(), so the way we insert intermediate node should not affect the output graph as original collective node will be removed anyway.

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 5 above)
Comment (fegin) [inline torch/_inductor/fx_passes/comm_fusion.py:68]: Can you add some comment here to mention that checking only one getitem is required because DDP fusion has exactly the same pattern except DDP fusion has more than one geittem?

**Comment by fegin** [comment]
cc., @yifuwang

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 1 above)
Comment (yifuwang) [reply torch/_inductor/fx_passes/comm_fusion.py:89]: Hey @yoyoyocmu, we'll soon migrate to a different set of aten-level collective IRs, which replace `(tag, ranks)` with `group_name` in the dispatcher op signature (see details here: https://github.com/pytorch/pytorch/issues/93173#issuecomment-1907095208). 

It should be trivial to adapt this pass once the migration happens. Just a heads up.

**Comment by yifuwang** [comment]
There seem to be a risk of de-syncing. Consider this valid SPMD program where the `getitem` index is not the rank:
```python
foo = all_reduce(...)
bar = split(foo, world_size)
baz = bar[0]
```

On rank 0, the program would be transformed into:
```python
baz = reduce_scatter(...)
```

However, on rank 1, the program would not be transformed, causing it to issue `all_reduce` when rank 0 would issue `reduce_scatter`.

I think there lacks a piece of infrastructure in Inductor for ensuring the robustness of this kind of optimizations. cc @Chillee

**Comment by yoyoyocmu** [comment]
Agree with @yifuwang that for these kind of collective communication fusion, the absence of a global view poses a risk of de-syncing. One potential scenario involves the replacement of only some ranks in the process group, while others retain the original function.

A prior discussion with @xw285cornell and @bdhirsh centered around an internal use case, where using DTensor automatically inserts `AllReduce() + Split()` in the backward propagation rule instead of `ReduceScatter()`.  Have two preliminary proposal to unblock this specific use case:
    1. Keep this fusion rule, allow users to annotate and mark specific collectives for optimization. If `torch.compile()` fails to fuse, throw an exception. This empowers users to control the behavior and take responsibility for the consequences.
    2. Enhance support in DTensor propagation rules to facilitate direct placement overwrite. This approach eliminates the need for this specific fusion in the compiler. cc @wanchaol 

Additionally, it appears that de-syncing is a common concern in the context of distributed training. Eager to gather more feedback from compiler experts and explore if there's a more robust and generic solution that can be devised to address this issue.

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 1 above)
Comment (xw285cornell) [reply torch/_inductor/fx_passes/comm_fusion.py:89]: I think @wconstab has a great point: torch.distributed.get_rank() gives you the rank id of the global PG. We should probably get the PG used by the allreduce. Otherwise I'd expect this won't work? (if you have 2D parallelism and the EP is in the inner dimension)

File: torch/_inductor/config.py (hunk 2 above)
Comment (yf225) [reply torch/_inductor/config.py:174]: +1 on making the flag name more specific.

In addition I think making it a list could be a good idea because it enforces an order with which we apply the passes (e.g. if user set `A_flag=True` and `B_flag=True`, it's not clear in which order they should be applied; but if user set `passes=[A, B]`, it's more clear that A should be applied before B).

For composability between passes, maybe we can test some common combinations and rely on user reporting issues on their models for edge cases?

**Comment by yf225** [comment]
also briefly discussed with @yifuwang on this: I believe we should categorize passes as "safe" and "unsafe", and:

1) ensure that important "safe" passes compose well together.
2) for "unsafe" passes (e.g. this one) that don't universally apply in all scenarios, we rely on the user to know which passes to choose and check accuracy on their own.

this way, when there is performance or accuracy issue, we will know to do ablation study on the enabled "unsafe" passes first, and then on the enabled "safe" passes.

**Comment by bdhirsh** [comment]
One thing I wonder if it’s worth doing: figure out how to trace get_rank() symbolically in the graph (like a dynamic shape / SymInt, but in a special way where we know it’s source comes from the get_rank API)

If we can do that, I think it would alleviate some of the safety concerns above. e.g we would know to pattern match on “baz = bar[get_rank()]” but not on “baz = bar[0]”.

One heavy-weight way to do this would be to introduce “symbolic rank” as a first class concept. A lighter-weight way would be to treat the output of “get_rank() as a plain symint, but plumb it’s source information to inductor.

**Comment by yifuwang** [comment]
@bdhirsh, that's a cool idea. I've also wondered about this. It seems that it's quite challenging to cover all valid sources for ranks (e.g., dist.get_rank(), os.environ["RANK"]. It's also valid to hard code ranks for broadcast and p2p operations). I'm also inclined to believe that such rank-specific optimizations are rather rare. IMO a more practical safety measure would be for the ranks to reach consensus on such optimizations.

For now, I agree with @yf225 on making such optimizations "use at your own risk", opt-in passes.

**Comment by bdhirsh** [comment]
> For now, I agree with @yf225 on making such optimizations "use at your own risk", opt-in passes.

Sounds good!

> It seems that it's quite challenging to cover all valid sources for ranks (e.g., dist.get_rank(), os.environ["RANK"]. It's also valid to hard code ranks for broadcast and p2p operations). I'm also inclined to believe that such rank-specific optimizations are rather rare.

Hmm - longer term, to me this feels like we have to decide which side to err on (performance vs. safety):

(a) be overly aggressive with this pattern match in a way that is sometimes unsafe (so we lose safety, which might be a problem if there are more of these unsafe optimizations in the future)
(b) don't be aggressive enough, and sometimes fail to match the pattern even when we should match it (so we lose performance)

I agree that covering all of the cases will be difficult, and in some cases, pretty much impossible. If a user assigned `x = get_rank()` to a variable in eager mode, and passes `x` into a compiled region, then we have no way to track the source properly. The user would pretty much have to re-write their code to call `get_rank()` inside of the compiled region. Although it would probably be possible to warn on situations where we "almost" match the pattern but fail to, as a kind of perf hint.

> IMO a more practical safety measure would be for the ranks to reach consensus on such optimizations.

Hmm this makes sense. Although - if the ranks don't reach consensus then I guess you would be forced to turn off the optimization completely? (unless we figure out how to make the patterns never match when they shouldn't).

File: torch/_inductor/fx_passes/comm_fusion.py (hunk 1 above)
Comment (yoyoyocmu) [reply torch/_inductor/fx_passes/comm_fusion.py]: The `rank_list` input of `all_reduce` is actually a list of global ranks (verified with 128 GPUs in the 2D case).

**Comment by Chillee** [comment]
@yifuwang I think the right way to handle this is that we should compile once for all ranks, and then re-use the graph on all ranks.
